pythonforandroid/bootstraps/common/build/jni/application/src/start.c

```diff
@@ -356,7 +358,91 @@ int main(int argc, char *argv[]) {
+    jstring j_python_path,
+    jstring j_arg) {
+  jboolean iscopy;
+  const char *android_private =
+      (*env)->GetStringUTFChars(env, j_android_private, &iscopy);
+  const char *android_argument =
+      (*env)->GetStringUTFChars(env, j_android_argument, &iscopy);
+  const char *service_entrypoint =
+      (*env)->GetStringUTFChars(env, j_service_entrypoint, &iscopy);
+  const char *python_name =
+      (*env)->GetStringUTFChars(env, j_python_name, &iscopy);
+  const char *python_home =
+      (*env)->GetStringUTFChars(env, j_python_home, &iscopy);
+  const char *python_path =
+      (*env)->GetStringUTFChars(env, j_python_path, &iscopy);
+  const char *arg = (*env)->GetStringUTFChars(env, j_arg, &iscopy);
+
+  setenv("ANDROID_PRIVATE", android_private, 1);
+  setenv("ANDROID_ARGUMENT", android_argument, 1);
+  setenv("ANDROID_APP_PATH", android_argument, 1);
+  setenv("ANDROID_ENTRYPOINT", service_entrypoint, 1);
+  setenv("PYTHONOPTIMIZE", "2", 1);
+  setenv("PYTHON_NAME", python_name, 1);
+  setenv("PYTHONHOME", python_home, 1);
+  setenv("PYTHONPATH", python_path, 1);
+  setenv("PYTHON_SERVICE_ARGUMENT", arg, 1);
+  setenv("P4A_BOOTSTRAP", bootstrap_name, 1);
+
+  char *argv[] = {"."};
+  /* ANDROID_ARGUMENT points to service subdir,
+   * so main() will run main.py from this dir
+   */
+  return main_(1, argv, 1);
+}
+
+#if defined(BOOTSTRAP_NAME_SERVICELIBRARY)
+// JNIEXPORT void JNICALL Java_org_kivy_android_PythonWorker_nativeStart(
+//     JNIEnv *env,
+//     jobject thiz,
+//     jstring j_android_private,
+//     jstring j_android_argument,
+//     jstring j_worker_entrypoint,
+//     jstring j_python_name,
+//     jstring j_python_home,
+//     jstring j_python_path) {
+//   jboolean iscopy;
+//   const char *android_private =
+//       (*env)->GetStringUTFChars(env, j_android_private, &iscopy);
+//   const char *android_argument =
+//       (*env)->GetStringUTFChars(env, j_android_argument, &iscopy);
+//   const char *worker_entrypoint =
+//       (*env)->GetStringUTFChars(env, j_worker_entrypoint, &iscopy);
+//   const char *python_name =
+//       (*env)->GetStringUTFChars(env, j_python_name, &iscopy);
+//   const char *python_home =
+//       (*env)->GetStringUTFChars(env, j_python_home, &iscopy);
+//   const char *python_path =
+//       (*env)->GetStringUTFChars(env, j_python_path, &iscopy);
+
+//   setenv("ANDROID_PRIVATE", android_private, 1);
+//   setenv("ANDROID_ARGUMENT", android_argument, 1);
+//   setenv("ANDROID_APP_PATH", android_argument, 1);
+//   setenv("ANDROID_ENTRYPOINT", worker_entrypoint, 1);
+//   setenv("PYTHONOPTIMIZE", "2", 1);
+//   setenv("PYTHON_NAME", python_name, 1);
+//   setenv("PYTHONHOME", python_home, 1);
+//   setenv("PYTHONPATH", python_path, 1);
+//   setenv("P4A_BOOTSTRAP", bootstrap_name, 1);
+
+//   char *argv[] = {"."};
+//   /* ANDROID_ARGUMENT points to service subdir,
+//    * so main() will run main.py from this dir
+//    */
+//   main_(1, argv, 0);
+// }
```

Comment:
Probably because it's still a WIP pull request, but I would not keep any commented out code in the final pull request

Reply:
Sure, i played around with running the python thread within the calling process (which is not the case for services). As long as no concurrent python gets started this works fine, but crashes otherwise. So this approach cannot be used when starting workers from e.g. kivy. It is required to use RemoteWorkerService/RemoteWorker in this case, which again uses a dedicated process. The commented function is a relict and will be removed during cleanup.